api/src/index.js

```diff
@@ -0,0 +1,217 @@
+        min: encoders[encName].autoOptimize.min,
+        max: encoders[encName].autoOptimize.max,
+        butteraugliDistanceGoal: optimizerButteraugliTarget,
+        maxRounds: maxOptimizerRounds,
+      },
+    );
+    binary = optimizedBinary;
+    optionsUsed = {
+      // 5 significant digits is enough
+      [optionToOptimize]: Math.round(quality * 10000) / 10000,
+    };
+  } else {
+    binary = encoder.encode(
+      bitmapIn.data.buffer,
+      bitmapIn.width,
+      bitmapIn.height,
+      encConfig,
+    );
+  }
+  return {
+    optionsUsed,
+    binary,
+    size: binary.length,
+  };
+}
+
+// both decoding and encoding go through the worker pool
+function handleJob(params) {
+  const { operation } = params;
+  switch (operation) {
+    case 'encode':
+      return encodeImage(params);
+    case 'decode':
+      return decodeFile(params);
+    case 'manipulate':
+      return manipulateImage(params);
+    default:
+      throw Error(`Invalid job "${operation}"`);
+  }
+}
+
+/**
+ * Represents an ingested image.
+ */
+class Image {
+  constructor (workerPool, file) {
+    this.workerPool = workerPool;
+    this.decoded = workerPool.dispatchJob({operation: 'decode', file});
+    this.encodedAs = {};
+  }
+
+  /**
+   * Define one or several manipulations to apply to the image.
+   * @param {object} manipulateOptions - An object with manipulations to apply, and their settings.
+   * @returns {Promise<undefined>} - A promise that resolves when all manipulations have completed.
+   */
+  async manipulate (manipulateOptions = {}) {
+    for (const [type, options] of Object.entries(manipulateOptions)) {
+      if (!Object.keys(manipulations).includes(type)) {
+        throw Error(`Invalid manipulation type "${type}"`);
+      }
+      const manipulatorOptions = Object.assign(
+        {},
+        manipulations[type].defaultOptions,
+        options,
+      );
+      this.decoded = this.workerPool.dispatchJob({
+        operation: 'manipulate',
+        manipulationType: type,
+        image: await this.decoded,
+        options: manipulatorOptions,
+      });
+      await this.decoded;
+    }
+  }
+
+  /**
+   * Define one or several encoders to use on the image.
+   * @param {object} encodeOptions - An object with encoders to use, and their settings.
+   * @returns {Promise<undefined>} - A promise that resolves when the image has been encoded with all the specified encoders.
+   */
+  async encode (encodeOptions = {}){
+    const { bitmap } = await this.decoded;
+    for (const [encName, options] of Object.entries(encodeOptions)) {
+      if (!Object.keys(encoders).includes(encName)) {
+        continue;
+      }
+      const encRef = encoders[encName];
+      const encConfig =
+        typeof options === 'string'
+          ? options
+          : Object.assign(
+              {},
+              encRef.defaultEncoderOptions,
+              options,
+            );
+      this.encodedAs[encRef.extension] = this.workerPool.dispatchJob({
```

Comment:
Hmm, that makes sense to me, since you might add more than one encoder that produces the same extension in the future.
I think the reason I decided to use the extension was that I found it very confusing how some of the encoders are just called 'webp' and 'avif', while others are called their actual name, such as 'mozjpeg', and so I didn't want to expose that if I could avoid it.

If we switch to the encoder name, I also think `encodedAs` should be renamed to `encodedWith`, and the returned object should include an `extension` string, so developers won't have to use a lookup table to figure out what extension they should use when they want to write the image to a file.

I'll wait for a LGTM from you before I make these changes.

Reply:
That make sense! Good shout!